drv: display: Initial FT800 display driver

hubertmis opened the pull request:
This patch includes initial support for FTDI FT800 display driver.
It includes basic features. It can be easily extended with more
FT800 display list commands and co-processor features.

Signed-off-by: Hubert Miś <hubert.mis@nordicsemi.no>

gmarull commented on drivers/display/ft8xx/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# FT8xx display controller configuration options
+
+# Copyright (c) 2020 Hubert Miś <hubert.mis@gmail.com>
```
Copyright should be consistent with commit author details (e-mail here)

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/display/ft8xx/ft8xx.h>
```
It looks like this is a local private header, so you don't need to use full path, just use `"ft8xx.h"`

hubertmis replied:
It's actually a public header placed in `include/drivers/display/ft8xx`. Should I remove some directories from the path? I thought that it is better to make include paths explicit.

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+#include <drivers/display/ft8xx/ft8xx.h>
+
+#include <stddef.h>
+#include <stdint.h>
```
I think these are not needed (e.g. stdint)

hubertmis replied:
`stddef.h` is needed to define `NULL` used in this module. `stdint.h` defines types like `uint8_t` and `uint32_t` used in this module. Are these headers includes indirectly and by convention not included in the driver modules?

gmarull replied:
these facilities are already provided when using Zephyr (e.g. stding types are used everywhere now)

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+
+#define DT_DRV_COMPAT ftdi_ft800
+
+#define LOG_MODULE_NAME ft8xx
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
```suggestion
LOG_MODULE_REGISTER(ft8xx, CONFIG_DISPLAY_LOG_LEVEL);
```

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+#include "ft8xx_host_commands.h"
+#include "ft8xx_memory.h"
+
+#define DT_DRV_COMPAT ftdi_ft800
```
this is usually defined at the top (convention)

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+#define HOFFSET   DT_INST_PROP(0, hoffset)
+#define HCYCLE    DT_INST_PROP(0, hcycle)
+#define HSYNC0    DT_INST_PROP(0, hsync0)
+#define HSYNC1    DT_INST_PROP(0, hsync1)
```
I think it's better to collect DT properties into driver config structure, see e.g.  https://github.com/zephyrproject-rtos/zephyr/pull/30960

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+
+#define FT8XX_EXPECTED_ID 0x7C
+
+static int_callback irq_callback;
```
Better use driver data to hold needed runtime variables

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+
+static void command(uint8_t cmd)
+{
+	(void)ft8xx_drv_command(cmd);
```
error chain lost here, better propagate errors

hubertmis replied:
Unfortunately FT800 API defines functions as `void`. Because of that it's hard to propagate errors. I've decided to just log errors to keep compatibility with API suggested in FT800 documentation.

gmarull replied:
`ft8xx_drv_command` is defined in this file and returns an error code if something goes wrong, so it should be propagated accordingly

hubertmis replied:
Please have a look at this snippet from FT800 Programmers Guide. 
![image](https://user-images.githubusercontent.com/21953319/103685789-161ece00-4f8e-11eb-9589-fd072f27bfdb.png)

Signature this function is expected to be `void host_command(uint8_t command)`. I'm not sure what action should be taken here if `ft8xx_drv_command()` returns an error. Assertion fault?

gmarull replied:
I've seen many examples/guides that do not take into account error handling, but real applications need to. If not, low level failures get buried resulting in hard to debug problems. As implementation is not external here, and you have the choice of adding error handling, I think it should be added.

hubertmis replied:
> you have the choice of adding error handling, I think it should be added

I understand your concern. However, creating own API for this device instead of following FT800 Programmers Guide would make hard porting applications using this display from other platforms. If the driver follows defined API, I can simply reuse display software modules from other projects. In similar way, I can reuse network modules using BSD socket API, because Zephyr follows defined socket API, instead of own improved implementation.

I think it would be better to decide on the best way of correct handling of low level errors inside the driver and keep API compatibility with the specification. My current proposal is to use error logs. I can implement something else, like a callback function called when operation fails.

What do you think?

gmarull replied:
In such case, then I'd place an assertion inside these functions so that when enabled the point of failure will be known.

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+	(void)ft8xx_drv_command(cmd);
+}
+
+static void busy_wait(void)
```
My understanding of a busy_wait is not a sleep call, so name may be misleading

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	/* Reset display controller */
+	command(CORERST);
```
error handling missing

hubertmis replied:
Unfortunately, FT800 API does not report errors

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+	busy_wait();
+
+	while (!verify_chip()) {
+		/* Intentionally empty */
```
can you explain why is this polling needed?

hubertmis replied:
It helps debugging hardware issues. If FT800 chip is incorrectly defined in device tree or incorrectly connected, the initialization procedure stops in this loop. I think it could be replaced with an error log

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+
+	/* Disable LCD */
+	wr8(REG_GPIO, 0);
+	wr8(REG_PCLK, 0);
```
error handling missing

hubertmis replied:
Like previously, FT800 functions like `command`, `wr8`, `wr32` does not return error codes. Error handling must be implemented in lower layers

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+	return 0;
+}
+
+SYS_DEVICE_DEFINE("ft8xx_spi", ft8xx_init, NULL,
```
Display drivers need to define a display API and register then via `DEVICE_DT_DEFINE` macro

hubertmis replied:
Like discussed below, FT800 does not expose display API. It defines its own API

gmarull commented on drivers/display/ft8xx/ft8xx.c:
```diff
@@ -0,0 +1,200 @@
+	/* Read REG_INT_FLAGS to clear IRQ */
+	uint8_t int_flags = rd8(REG_INT_FLAGS);
+	/* This list should contain only TAG interrupt */
+	(void)int_flags;
```
what is the purpose of this line?

hubertmis replied:
Suppressing compiler warning that this value is unused. It could be used in the future to determine what event caused an interrupt. Currently only TAG event is supported and because of that checking is not needed.

gmarull commented on drivers/display/ft8xx/ft8xx_common.c:
```diff
@@ -0,0 +1,91 @@
+static void encode_u16(uint8_t *buffer, uint16_t data)
+{
+	buffer[0] = (uint8_t)data;
+	buffer[1] = (uint8_t)(data >> 8);
```
You may take a look at `sys_cpu_to_XXXX` family of functions

gmarull reviewed (changes requested):
I've done an initial partial review, left a few comments. Some high level description would help on understanding the current code structure/approach, as it deviates from other display drivers in multiple aspects.

hubertmis commented:
Thanks @gmarull for the review. Before I answer to your comments, I would like to add overview of this PR.

> Some high level description would help on understanding the current code structure/approach, as it deviates from other display drivers in multiple aspects.

I've created this driver for a bare-metal project that I'm porting to Zephyr right now. I think that it could be useful for other Zephyr users.

I've reviewed display driver API in Zephyr and it looks that it is designed for display controller exposing a frame buffer interface. FT800 does not follow this approach. It exposes higher-layer features like drawing lines, bitmaps transformation, rendering text and numbers using fonts. Also, FT800 Programmers Guide suggests API to use with this controller. Because of that I have not used Zephyr display driver API at all, but created separated API following FT800 recommendations. I've added the driver to drivers/display directory, because it is a display controller driver. But it does not expose API defined in include/drivers/display.h. It exposes separated API defined in include/drivers/display/ft8xx/*

gmarull commented:
Thanks for the clarification @hubertmis. I don't think then that the driver should fall into the display area, as it doesn't offer a display API interface. I suppose a new category (i.e. API) should be created if there are enough devices like this one.

hubertmis commented:
Before we move forward with this review, I think we should decide where is the correct place for device drivers that expose higher-layer API than expected in Zephyr device driver interface for given type of device.

I think that examples of similar devices would be:
* 802.15.4 radio chip exposing MAC layer interface, while Zephyr expects closer to PHY layer,
* Bluetooth chip exposing host interface, while Zephyr expects controller.

I think that drivers for such devices should be implemented in `subsys/net/l2/` and `subsys/bluetooth/` accordingly. Such drivers would be placed alongside other 802.15.4 MAC layers, and Bluetooth Host implementations.

Because of that, probably good place for the driver of a higher-layer display controller like FT800, would be in a directory like `subsys/display/`. What do you think?

vanwinkeljan commented:
I had a quick look into to this device and it feels more like a GPU then a display controller, and maybe it should be handle as such.

hubertmis commented:
@vanwinkeljan, is there any GPU support in Zephyr? I could not find any indication in the source code or documentation.

vanwinkeljan commented:
There is no support yet for GPUs.

I'm just wondering if we should put the code in a sub-dir `gpu` instead of `display`.

On the `subsys` vs `drivers`, I still have the feeling that it belongs more in `drivers` as it deals with the hardware.
In the networking space there is `net_offload`, is this not something similar?

gmarull commented:
I'm not sure it can be considered as purely a GPU such as the IP contained by some high-end MCUs like STM32 or NXP (which offer hw acceleration for common 2D operations like copy, blend, etc.). It looks like a highly custom IC.

![image](https://user-images.githubusercontent.com/25011557/103346613-c7a49780-4a94-11eb-94b9-ed6911a10c75.png)

![image](https://user-images.githubusercontent.com/25011557/103346503-7399b300-4a94-11eb-8001-c7d1f5d5afdb.png)

hubertmis commented:
If I understand correctly, the idea for the graphics subsystem in Zephyr is like this:
1. The lowest layer is a display driver operating on a frame buffer (include/drivers/display.h).
2. The high layer is a graphics library like LVGL that exposes API for the application and uses display driver as a backend.
3. The actual graphic is created in the application using features exposed by the graphics library.

It looks that FT800 exposes similar features to 1 and 2 combined, however it uses other API than LVGL. Because of that it should be a design choice of application developer to use a flexible and portable library like LVGL, or to offload graphics creation to a separated chip like FT800. If I understand net offloading correctly, it works in different manner, because it shares the same API (BSD sockets), but offloads some operations to a separated device. FT800 requires different API.

I'm not sure if any of these layers can be considered as GPU. I understand GPU more like a generic processing unit that can be used with a generic library (Vulkan, CUDA).

Perhaps the devices like FT800 could be called "graphics engine" instead of display controller or GPU. However, I'm not aware of any other graphics engines that could be added to the same category. Perhaps a new directory like `drivers/misc` could be created to cover such one-of-a-kind devices that do not expose Zephyr driver API?

vanwinkeljan commented:
The `drivers/misc` solution does not sound bad, I wonder how others think about it (@nashif @MaureenHelm )?

Some other open points form my side:
1. Missing sample
2. Missing documentation

hubertmis commented:
@vanwinkeljan, I'm not sure how could I create a public sample if the only board with this chip I'm aware of is my custom PCB. Shall I create a sample and verify with my board? Such sample would be unbuildable by anyone else, unless one creates a DTS file adding FT800 to a project.

jfischer-no commented:
> The `drivers/misc` solution does not sound bad,

Disagree, putting it in a directory that has nothing to do with display is wrong. There are several similar devices from the vendor.
I think drivers/display/ft8xx/ should be more suitable for the whole family. However, this is the minor part of the whole PR.

> @vanwinkeljan, I'm not sure how could I create a public sample if the only board with this chip I'm aware of is my custom PCB. Shall I create a sample and verify with my board? Such sample would be unbuildable by anyone else, unless one creates a DTS file adding FT800 to a project.

VM800C for example is publicly available FT800 development module, you could add a generic Arduino shield for it.

hubertmis commented:
> VM800C for example is publicly available FT800 development module, you could add a generic Arduino shield for it.

Unfortunately, I don't own this module and I'm not planning to buy it. Could a sample be created later, by someone who can verify with this module?

jfischer-no commented:
> > VM800C for example is publicly available FT800 development module, you could add a generic Arduino shield for it.
> 
> Unfortunately, I don't own this module and I'm not planning to buy it. Could a sample be created later, by someone who can verify with this module?

The driver alone is not enough, you need something CI can build on and reviewer to see how it is used. The best thing you can do is to add VM800C shield support, even you do not own it. Much of the work around shield support is needed anyway, for example bindings that describe the display panel, which would be a child of the FT8xx controller node.

hubertmis commented:
> The best thing you can do is to add VM800C shield support, even you do not own it.

OK, I'll do that.

Before I push my changes I would like to establish to which directory add the driver to avoid moving it around. I'll push changes after it is decided.

gmarull commented:
> Before I push my changes I would like to establish to which directory add the driver to avoid moving it around. I'll push changes after it is decided.

IMO if driver can't expose a display API interface it should not fall into the display category. It's not only a display controller, it seems to have a touch controller, a graphics co-processor... so `drivers/display` doesn't seem to be the right place. I think that these kind of devices are hard (if possible) to abstract, so they should be used directly by the application using a custom API (could be setup as a Zephyr module?)

jfischer-no commented:
> IMO if driver can't expose a display API interface it should not fall into the display category.

Disagree, we already have display or display/-controller drivers in _driver/display_ that do not use the display API and they were there before Display API.  And the fact is that the FT800 drives a display panel.

hubertmis commented:
I've implemented review suggestions, and added a shield and a sample.
I kept the driver inside `drivers/display`, because there is already `grove_lcd_rgb` that works in a similar way (does not expose display API, but defines own).

There is an important limitation of FT800 API and this driver inherits this limitation: only one instance of FT800 is supported by the system.

carlescufi commented:
@jfischer-no could you take another look please?

gmarull reviewed (changes requested):
A couple of general comments:

- Changes should be splitted in multiple commits (e.g. driver and sample)
- I'm still not comfortable of making this driver part of `drivers/display` as FT800 is not only a display controller, but also has touch, audio, etc. So maybe a sort of "misc" category should be created with drivers that have custom APIs.

hubertmis committed:
drv: display: Initial FT800 display driver

This patch includes initial support for FT800 display driver.
It includes basic features. It can be easily extended with more
FT800 display list and co-processor features.

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

hubertmis committed:
shields: FTDI VM800C

This patch introduces support for FTDI VM800C including FT800
Embedded Video Engine

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

hubertmis committed:
samples: FT800 demo application

This patch adds a new sample application demonstrating basic
usage of FT800 display controller including touchscreen events.

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

MaureenHelm commented:
> > * I'm still not comfortable of making this driver part of `drivers/display` as FT800 is not only a display controller, but also has touch, audio, etc. So maybe a sort of "misc" category should be created with drivers that have custom APIs.
> 
> I think it would be cleaner if drivers not exposing Zephyr APIs were moved out of `drivers/display` to other directory like `misc`. It applies to FT800 and Grove LCD. I'm happy to move FT800 driver there. But I think it should be decided between Zephyr developers first. It seems there are conflicting opinions.

@carlescufi @galak Let's discuss this in the API or dev-review meeting

gmarull commented:
> 
> 
> 
> 
> > > * I'm still not comfortable of making this driver part of `drivers/display` as FT800 is not only a display controller, but also has touch, audio, etc. So maybe a sort of "misc" category should be created with drivers that have custom APIs.
> 
> > 
> 
> > I think it would be cleaner if drivers not exposing Zephyr APIs were moved out of `drivers/display` to other directory like `misc`. It applies to FT800 and Grove LCD. I'm happy to move FT800 driver there. But I think it should be decided between Zephyr developers first. It seems there are conflicting opinions.
> 
> 
> 
> @carlescufi @galak Let's discuss this in the API or dev-review meeting

Sounds good to discuss that in either API/dev-review meetings. I'll try to join.

hubertmis commented:
I've modified directory structure according to solution proposed during API meeting.

I have not moved `grove_lcd.h` or `mb_display.h` to the newly created directory. These files should be moved in a separated PR.

hubertmis commented:
Are there any recommendations who should be code owner of `drivers/misc` directory and `drivers/misc/ft8xx` driver?

carlescufi commented:
> Are there any recommendations who should be code owner of `drivers/misc` directory and `drivers/misc/ft8xx` driver?

Do not add an owner for `drivers/misc`, but add yourself for `drivers/misc/ft8xx`.

hubertmis commented:
> Do not add an owner for `drivers/misc`

Compliance complains about files in `drivers/misc` without owner:

```
drivers/misc/CMakeLists.txt
drivers/misc/Kconfig
```

carlescufi commented:
> > Do not add an owner for `drivers/misc`
> 
> Compliance complains about files in `drivers/misc` without owner:
> 
> ```
> drivers/misc/CMakeLists.txt
> drivers/misc/Kconfig
> ```

OK, add @tejlmand  then because it's really just build system.

carlescufi reviewed (changes requested):
`samples/subsys/display/ft8xx` should really be `samples/drivers/misc/ft8xx`

gmarull commented on drivers/misc/ft8xx/ft8xx_drv.c:
```diff
@@ -0,0 +1,202 @@
+
+static struct gpio_callback irq_cb_data;
+
+__WEAK void ft8xx_drv_irq_triggered(const struct device *dev,
```
doesn't zephyr define `__weak` (lowercase)?

gmarull commented on drivers/misc/ft8xx/ft8xx_drv.c:
```diff
@@ -0,0 +1,202 @@
+
+	spi_ft8xx_dev = device_get_binding(DT_BUS_LABEL(NODE_ID));
+	if (!spi_ft8xx_dev) {
+		return -EINVAL;
```
`-ENODEV`

gmarull commented on drivers/misc/ft8xx/ft8xx_drv.c:
```diff
@@ -0,0 +1,202 @@
+	 */
+	int_ft8xx_dev = device_get_binding(DT_GPIO_LABEL(NODE_ID, irq_gpios));
+	if (!int_ft8xx_dev) {
+		return -EINVAL;
```
`-ENODEV`

gmarull commented on drivers/misc/ft8xx/ft8xx_drv.h:
```diff
@@ -0,0 +1,36 @@
+#include <stdint.h>
+
+#include <drivers/gpio.h>
+#include <zephyr.h>
```
you can likely just include `device.h` here instead of `zephyr.h`

gmarull commented on include/drivers/misc/README:
```diff
@@ -0,0 +1,17 @@
+Miscellaneous Drivers
```
I think that this information should go somewhere down to `doc/reference`

hubertmis replied:
Proposed solution from API meeting recommended to place it here:

>Add a include/drivers/misc/README to clarify what goes in there. In that README, there should be a description of each item in the folder, with an explanation of why it doesn't conform to the standard API.

gmarull replied:
ok, fair enough. Could it be `.rst` and have some minimal formatting then?

hubertmis replied:
The proposal was to put `README`. I'm not sure if omitting `.rst` or `.md` was intentional. @carlescufi, please comment

gmarull commented on include/drivers/misc/ft8xx/ft8xx.h:
```diff
@@ -0,0 +1,42 @@
+extern "C" {
+#endif
+
+struct ft8xx_touch_transform {
```
public API should be documented (data structures, functions, etc.). there should be a new doxygen group for misc drivers and place each driver into its own group.

hubertmis commented on drivers/misc/ft8xx/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config FT800
+	bool "FT800 Embedded Video Engine driver"
+	select SPI
```
Do I understand correctly that replacing this entry with `depends on SPI` will make `FT800` option invisible, unless one enables `SPI`?

Is it what we want to achieve here?

gmarull replied:
Yes, this is what most Zephyr drivers do. See https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#select-recommendations

gmarull reviewed (approved):
LGTM. I'd prefer to make the readme a rst file so that it is displayed better when browsed on GH.

hubertmis commented:
Is it possible to somehow preview doxygen docs? I don't see it in documentation build artifacts. I would like to make sure it renders as I expected, because I'm not fluent in doxygen syntax.

jfischer-no commented on include/drivers/misc/ft8xx/ft8xx_common.h:
```diff
@@ -0,0 +1,99 @@
+/** Short notation to write 2 bytes to FT8xx memory. */
+#define wr16 ft8xx_wr16
+/** Short notation to write 4 bytes to FT8xx memory. */
+#define wr32 ft8xx_wr32
```
Please make the whole defines and wrappers away and use the functions directly (also all the encode and parse wrappers).

hubertmis replied:
These wrappers are needed to align with API defined in FT800 Programmer Guide:
![image](https://user-images.githubusercontent.com/21953319/111278943-6ffcdd80-863a-11eb-97a3-fb89e7291362.png)

jfischer-no replied:
How that applies to this Zephyr OS driver or why do you have to follow it? What am I missing?

hubertmis replied:
It's like POSIX API for sockets. There are multiple libraries using POSIX API and now you can easily port those libraries to Zephyr, because Zephyr exposes widely used POSIX socket API.

In similar way, you with this PR you can easily port libraries using FT800 to Zephyr applications, because this driver exposes widely used public API defined in FT800 Programmer Guide.

If I removed these macros, I would need to rewrite each library using FT800 I would like to use in Zephyr-based projects.

jfischer-no replied:
okay, I see no reason to copy this very bad example. This "API" must have a namespace. Even if this "API"  would have proper namespace prefix it should not be done using macros but inline functions.

> If I removed these macros, I would need to rewrite each library using FT800 I would like to use in Zephyr-based projects.

Well it is not that much work to grep and prefix all the calls with "ft8xx_" or  you could just add a wrapper in your application code.

hubertmis replied:
I can easily change macros to inline functions, but I would like to keep these in Zephyr driver. Moving it to each library integration code is more complicated for me, than maintaining my own FT800 driver for Zephyr in project repositories.

I see no point in sharing my driver with community if I cannot easily use it in my projects.

gmarull replied:
I think that if there are a lot of existing FT800 examples using this "API" we should just provide it to ease portability. However, because it is easy to have name clashes with such short names, we should probably restrict its usage to external samples and not code in Zephyr.

hubertmis replied:
> it is easy to have name clashes with such short names

That's exactly why I designed Zephyr driver in a way that public symbols have `ft8xx_` prefix, to avoid name collisions. Only compilation units including `ft8xx_common.h`, `ft8xx_copro.h`, or `ft8xx_memory.h` can be subjected to such collisions. Such compilation units are usually responsible for display, touch processing, or audio and do not include other libraries that may be prone to clashes.

Perhaps I could create separated header like `ft8xx_reference_api.h` including only `static inline` mappings of `ft8xx_...()` functions to shorter names. This way we would avoid name conflicts, because only compilation units including this header would be troubled. And there would be no other reason to include such file than to use this defined API.

What do you think?

gmarull replied:
The problem may appear if Zephyr ever exposes such names via a sufficiently common header. I think the best we can do is to restrict such helpers to external applications (maybe sample at most for now).

jfischer-no replied:
> Moving it to each library integration code is more complicated for me, than maintaining my own FT800 driver for Zephyr in project repositories.
> 
> I see no point in sharing my driver with community if I cannot easily use it in my projects.

I do not see how this PR can go in if the public driver specific "API" does not use a namespace, regardless via what public header it is exposed. Others can argue just like you, why should they not get the right to use the same identifiers? How should this drivers work together in an application?

hubertmis replied:
> How should this drivers work together in an application?

Obviously if you use multiple poorly designed APIs (like `wr32()`) inside single compilation unit, it cannot work correctly. However, this driver does not expose any global symbols without "namespace" prefixes. These short names are affecting only compilation units that include header files from this driver.

I agree with @gmarull that we should limit these includes in Zephyr to only samples to make sure that these files are not included indirectly.

I'll prepare a proposal in the evening separating these non-namespace from the rest of the code. I hope such solution will be acceptable by everyone. We can discuss later after I push it.

hubertmis replied:
I've moved not-prefixed names from `ft8xx_common.h` and `ft8xx_copro.h` to a new header file `ft8xx_reference_api.h`. This new header file is included only in the FT800 sample (and in platform integration in my projects). This way only single sample main.c compilation unit is polluted with this potentially conflicting API. Problematic header cannot be unintentionally indirectly included, because none of headers include it. I've also added a note to the documentation of this header to use it with caution because of potential naming conflicts.

jfischer-no replied:
This sample is also what I am talking about, something like `cmd(CLEAR_COLOR_RGB(0x00, 0x00, 0x00));` implies `cmd()` is some common API but it is not. This sample should only use the prefixed driver's specific API, ft8xxx_cmd() and not cmd(). People will use it and take this bad example for their own application or even worse for another driver.

hubertmis replied:
I agree that we should show using `ft8xx_` prefixed names as an example. I'm only concerned that no code in Zephyr includes `ft8xx_reference_api.h` now and because of this it's not easy to even verify if syntax in this file is correct using a compiler.

However, this file is relatively simple and I checked it with an out-of-tree application. It's not a big problem to maintain it without using in Zephyr directly.

hubertmis commented on drivers/misc/ft8xx/ft8xx_common.c:
```diff
@@ -0,0 +1,91 @@
+	int err;
+	uint8_t buffer[2];
+
+	encode_u16(buffer, data);
```
Good point. I can use it directly here. I ported this file from other platform where abstraction lie `sys_cpu_to...()` was not present

carlescufi reviewed (approved):
LGTM now, thank you!

gmarull commented:
> Is it possible to somehow preview doxygen docs? I don't see it in documentation build artifacts. I would like to make sure it renders as I expected, because I'm not fluent in doxygen syntax.

You can do this from `doc` dir:
```
cmake -GNinja -B_build .
ninja -C _build doxygen
```

gmarull commented on include/drivers/misc/ft8xx/ft8xx_memory.h:
```diff
@@ -0,0 +1,236 @@
+#if defined(CONFIG_FT800)
+/** Main parts of FT800 memory map */
+enum ft8xx_memory_map_t {
+	RAM_G         = 0x000000,
```
I think all these enums should be prefixed, ie `FT8XX_`

hubertmis replied:
I prefixed with `FT800_` and `FT810_`, because memory map differs between these devices.

gmarull commented on include/drivers/misc/ft8xx/ft8xx_dl.h:
```diff
@@ -0,0 +1,225 @@
+ */
+
+/** Rectangular pixel arrays, in various color formats */
+#define BITMAPS      1U
```
names in this file may also be prefixed with `FT8XX_`

gmarull commented on include/drivers/misc/ft8xx/ft8xx_copro.h:
```diff
@@ -0,0 +1,148 @@
+ */
+
+/** Co-processor widget is drawn in 3D effect */
+#define OPT_3D        0
```
definitions in this file may be prefixed with `FT8XX_`

gmarull reviewed (commented):
IMO It's ok to have `ft8xx_reference_api.h ` file with advice. Anyone using such "API" is at least warned. Some other public headers contain non prefixed definitions, I think they should unless they belong to the "reference API".

gmarull reviewed (approved):
Current solution LGTM. The "reference API" is isolated enough and up to the application to use it or not.

gmarull commented:
@hubertmis `FT8XX_REG_TOUCH_TAG` is causing some doc errors
